### PR TITLE
Deprecate iris.proxy

### LINKED ISSUE #1838: Deprecate iris.proxy
As requested : https://github.com/SciTools/iris/pull/1789#discussion_r40073480

I've investigated this and will post a PR shortly..

### PR DESCRIPTION
**Pull request by rhattersley**
Closes #1838.

### CODE REVIEW

**Comment by rhattersley** [comment]
@pp-mo - Travis-CI has given this the :+1:

**Comment by rhattersley** [comment]
Pinging @pelson since you [raised](https://github.com/SciTools/iris/pull/1789#discussion_r40073480) the issue in the first place.

File: lib/iris/fileformats/netcdf.py [hunk 1]
```diff
@@ -35,8 +35,7 @@
 import warnings
 
 import biggus
-import iris.proxy
-iris.proxy.apply_proxy('netCDF4', globals())
+import netCDF4
```
Comment (pelson) [inline lib/iris/fileformats/netcdf.py]: netCDF isn't an optional dependency is it?

Comment (pp-mo) [inline lib/iris/fileformats/netcdf.py]: Surprisingly, I think we are "close" to this being an optional dependency.
Apart from the obvious fileformat modules we have no direct I think there are no essential dependencies on NetCDF4.

However, a lot of things do need netcdftime.
That means a lot of *other* file formats (like PP and ff-fast-structured).  There are some uses of this in the core modules too : coords, cube, unit (though that is now going), and util.
(note that 'util' is also core as other core modules like coords and cube import it).

Comment (pp-mo) [inline lib/iris/fileformats/netcdf.py]: > close to optional

Does that affect our attitude to this change, though ?
Probably not.

**Comment by pp-mo** [comment]
@rhattersley @pelson I'm tempted to just merge this, as I don't see a real downside.
Any more objections ?

**Comment by QuLogic** [comment]
On a related note, the change in #1842 allows docs to build with missing dependencies, so the reason for `iris.proxy` noted in the comments is no longer true.

File: lib/iris/experimental/regrid_conservative.py [hunk 2]
```diff
@@ -76,6 +72,9 @@ def _make_esmpy_field(x_coord, y_coord, ref_name='field',
         Add a mask item to the grid, assigning it 0/1 where mask=False/True.
 
     """
+    # Lazy import so we can build the docs with no ESMF.
```
Comment (pelson) [inline lib/iris/experimental/regrid_conservative.py]: Perhaps the comment should be "Lazily import ESMF, as it is only an optional dependency.".

File: lib/iris/fileformats/netcdf.py (hunk 1 above)
Comment (rhattersley) [inline lib/iris/fileformats/netcdf.py]: `netCDF4` isn't optional (e.g. we list it in `minimal-conda-requirements.txt`), hence I've just made it a regular import.

**Comment by rhattersley** [comment]
> On a related note, the change in #1842 allows docs to build with missing dependencies, so the reason for iris.proxy noted in the comments is no longer true.

The change in #1842 isn't equivalent. #1842 gives a stub page, but a lazy import (or iris.proxy) allows the full documentation to be built.

File: lib/iris/experimental/regrid_conservative.py (hunk 2 above)
Comment (rhattersley) [inline lib/iris/experimental/regrid_conservative.py]: As noted below, the lazy import is only really needed so we can build full documentation without needing the dependency.

**Pull request merged**